apps/alerts/lib/cache/bus_stop_change_s3.ex

```diff
@@ -0,0 +1,119 @@
+  @moduledoc """
+  Functionality for writing to/reading from S3. Takes %Alert{} converts them into %HistoricalAlert{} before saving.
+  """
+  use RepoCache, ttl: :timer.hours(24)
+  require Logger
+  alias Alerts.{Alert, HistoricalAlert}
+
+  @ex_aws Application.compile_env(:alerts, [:aws_client], ExAws)
+  @ex_aws_s3 Application.compile_env(:alerts, [:s3_client], ExAws.S3)
+  @bucket "mbta-dotcom"
+  @bucket_prefix "bus-stop-change"
+
+  @doc """
+  Arguments for the Alerts.Cache.Fetcher process, Fetches current bus alerts
+  hourly, sending the parsed alerts to copy_alerts_to_s3/2
+  """
+  def fetcher_opts do
+    [
+      api_mfa:
+        {V3Api.Alerts, :all,
+         [
+           [
+             "filter[route_type]": 3,
+             "filter[datetime]": "NOW"
+           ]
+         ]},
+      repeat_ms: 3_600_000,
+      update_fn: &copy_alerts_to_s3/2
+    ]
+  end
+
+  @doc """
+   Find alerts that qualify as a bus stop closure or
+   stop move, and copy them to our S3 bucket. This function is invoked in Alerts.Cache.Fetcher.handle_info/2 with two arguments, but we only use the first.
+  """
+  @spec copy_alerts_to_s3([%Alert{}], any) :: :ok
+  def copy_alerts_to_s3(bus_alerts, _) do
+    with stop_change_alerts when stop_change_alerts != [] <-
+           Enum.filter(bus_alerts, &(&1.effect in [:stop_closure, :stop_moved])) do
+      Enum.map(stop_change_alerts, & &1.id)
+      |> Enum.sort()
+      |> cache(fn _sorted_alert_ids ->
+        stop_change_alerts
+        |> Enum.map(&HistoricalAlert.from_alert/1)
+        |> write_alerts()
+      end)
+    else
+      [] ->
+        # No stop change alerts to write.
+        :ok
+    end
+  end
+
+  @doc """
+  Get bus stop change alerts currently stored on S3.
+  """
+  @spec get_stored_alerts :: [HistoricalAlert.t()]
+  def get_stored_alerts do
+    keys =
+      @ex_aws_s3.list_objects(@bucket, prefix: @bucket_prefix)
+      |> @ex_aws.stream!
+      |> Stream.map(& &1.key)
+      |> Enum.to_list()
+
+    Enum.map(keys, fn key ->
+      result =
+        @ex_aws_s3.get_object(@bucket, key, prefix: @bucket_prefix)
+        |> @ex_aws.request()
+
+      case result do
+        {:ok, %{body: alert_data}} ->
+          decompress_alert(alert_data)
+
+        error ->
+          log_result(error, key, "get_stored_alerts")
+          nil
+      end
+    end)
+    |> Enum.filter(& &1)
+  end
+
+  @spec write_alerts([HistoricalAlert.t()]) :: :ok
+  defp write_alerts(alerts) do
+    alerts
+    |> Enum.map(fn alert ->
+      {alert.id, compress_alert(alert)}
+    end)
+    |> Task.async_stream(
+      fn {id, contents} ->
+        @ex_aws_s3.put_object(@bucket, "#{@bucket_prefix}/#{id}", contents)
+        |> @ex_aws.request()
+        |> log_result(id, "write_alerts")
+      end,
+      max_concurrency: 10
+    )
+    |> Stream.run()
+  end
+
+  @spec compress_alert(HistoricalAlert.t()) :: binary()
+  defp compress_alert(alert) do
+    :erlang.term_to_binary(alert, [:compressed])
```

Comment:
Since I'm _only_ using these directly in a template with various functions that operate on an `%Alert{}`, it didn't feel worthwhile to add an extra JSON serializing/deserializing step I guess? :)